185454357 update mci client job

ZachCMP opened the pull request:
https://www.pivotaltracker.com/story/show/185454357

ZachCMP committed:
Add UpdateMciClientJob

ZachCMP committed:
Run job when client is updated

ZachCMP commented:
@gigxz I took my best shot at what this is supposed to look like. I can go add tests if you like, but I don't think I have enough context about the expected behavior on the MCI side to be able to write them quickly. Let me know if you want me to give it a shot regardless.

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -167,6 +168,22 @@ def mci_id
     ac_hmis_mci_ids.to_a.min_by(&:id)&.value
   end
 
+  def update_mci_client
```
Hmm we should only do this on form submission, not every time the client is saved. I don't think there is a "save context" like there is for validation contexts, so I think the only way is to set an attribute like we have above.

```
    # Post-save action to create a new MCI ID if specified by the ClientProcessor
    if update_mci_attributes && HmisExternalApis::AcHmis::Mci.enabled?
      self.update_mci_attributes = nil
      # run job if relevant attributes changed
    end
```

Then in the ClientProcessor, we should set that attribute in [this case](https://github.com/greenriver/hmis-warehouse/blob/5d70e6ce358172f23c9fc205472e1d8914106b15/drivers/hmis/app/models/hmis/hud/processors/client_processor.rb#L160-L161) and [if we get here](https://github.com/greenriver/hmis-warehouse/blob/5d70e6ce358172f23c9fc205472e1d8914106b15/drivers/hmis/app/models/hmis/hud/processors/client_processor.rb#L184).

Does that make sense?

ZachCMP replied:
Do you mean that it should be set to `false` in the former case and `true` in the latter? That makes sense if so. It that's not what you mean, can you please explain in more detail?

gigxz replied:
In both cases I linked, it should be set to "true".

The first case is when someone is submitting a form for a client that has an MCI ID. When that happens, we want to hit their API if any of the relevant fields have changed, so that they are aware of any new data associated with that ID.
The second case is when someone is linking a client to an MCI ID for the first time. When that happens, we want to hit their API with the relevant fields, so that they are aware of any new data associated to that ID. It's fine to only make the call if any of the MCI-relevant fields have `changed?` because they will have changed if the MCI ID is added.

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -167,6 +168,22 @@ def mci_id
   end
 
+  def update_mci_client
+    return unless ac_hmis_mci_ids.exists?
```
If you implement my suggestion above, we can remove this. We don't want to run this query every time the client is saved. The job will throw an error if the client doesnt have an MCI ID. (which shouldn't happen if we only invoke this actio based on a `update_mci_attributes` attribute).

gigxz commented on drivers/hmis/app/models/hmis/hud/client.rb:
```diff
@@ -167,6 +168,22 @@ def mci_id
+      'GenderOther',
+    ].any? { |field| previous_changes&.[](field) }
+
+    HmisExternalApis::AcHmis::UpdateMciClientJob.perform_later
```
Don't we need to pass the client?
```suggestion
    HmisExternalApis::AcHmis::UpdateMciClientJob.perform_later(client_id: id)
```

ZachCMP committed:
Merge remote-tracking branch 'origin/release-80' into 185454357-update-mci-client-job

ZachCMP committed:
PR Feedback

ZachCMP commented:
@gigxz Addressed your feedback. I did what I think you meant in the first comment, but see my question for more info. Let me know if you need anything else on this

gigxz commented:
Thanks! looks like there are some test failures fyi

ZachCMP committed:
Fix tests

ZachCMP commented:
@gigxz Test failures are fixed. Looks like brakeman is complaining about a CVE unrelated to this code. Let me know if you want me to do anything about that, or if you need anything else here.

eanders commented:
CVE is fixed in the base branch

gigxz committed:
Merge branch 'release-80' into 185454357-update-mci-client-job

gigxz committed:
Comments and reorg for clarity, fix bad merge

gigxz committed:
Comments and reorg for clarity, fix bad merge

gigxz committed:
Set userID on payload, commented-out source syskey

gigxz committed:
Merge branch 'release-80' into 185454357-update-mci-client-job

gigxz commented:
There seems to be an issue with updateclient endpoint not accepting any `userID`, so I haven't gotten a successful run on this yet. Will merge anyway, I think the issue is on their end.

gigxz reviewed (approved):
thanks @ZachCMP this is great

gigxz committed:
Fix unexpected return
